cartographer/io/mapping_state_deserializer_test.cc

```diff
@@ -0,0 +1,179 @@
+  return msg;
+}
+
+template <typename T>
+std::unique_ptr<T> CreateProtoUPtrFromString(const std::string& proto_string) {
+  return make_unique<T>(ProtoFromStringOrDie<T>(proto_string));
+}
+
+// open text file with a set of text-protos that mimic a pbstream.
+std::unique_ptr<InMemoryProtoStreamReader>
+CreateInMemoryReaderFromTextProtos() {
+  std::queue<std::unique_ptr<Message>> proto_queue;
+
+  proto_queue.push(
+      CreateProtoUPtrFromString<mapping::proto::SerializationHeader>(
+          testing::kSerializationHeaderProtoString));
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kPoseGraphProtoString,
+                       proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kAllTrajectoryBuilderOptionsProtoString,
+                       proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kSubmapProtoString, proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kNodeProtoString, proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kTrajectoryDataProtoString,
+                       proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kImuDataProtoString, proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kOdometryDataProtoString,
+                       proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kFixedFramePoseDataProtoString,
+                       proto_queue.back().get());
+
+  proto_queue.push(make_unique<mapping::proto::SerializedData>());
+  ProtoFromStringOrDie(testing::kLandmarkDataProtoString,
+                       proto_queue.back().get());
+
+  return make_unique<InMemoryProtoStreamReader>(std::move(proto_queue));
+}
+
+// This test checks, if the serialization works.
+TEST(MappingStateDeserializerTest, WorksOnGoldenTextStream) {
+  // Load text proto into in_memory_reader.
+  std::unique_ptr<InMemoryProtoStreamReader> reader =
+      CreateInMemoryReaderFromTextProtos();
+
+  io::MappingStateDeserializer deserializer(reader.get());
+
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
+      deserializer.header(), ProtoFromStringOrDie<SerializationHeader>(
+                                 testing::kSerializationHeaderProtoString)));
+
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
+      deserializer.pose_graph(),
+      ProtoFromStringOrDie<SerializedData>(testing::kPoseGraphProtoString)
+          .pose_graph()));
+
+  EXPECT_TRUE(google::protobuf::util::MessageDifferencer::Equals(
```

Comment:
You could use EXPECT_EQ and a matcher.
Then, a failed test would print the two arguments.

Reply:
added TODO